src/MessagePack.UnityClient/Assets/Scripts/MessagePack/Resolvers/CompositeResolver.cs

```diff
@@ -38,124 +30,102 @@ public CompositeResolver()
         /// The composite resolver that includes the AOT code generated resolver and several standard ones
         /// that the AOT code generator expects to be present.
         /// </returns>
-        public static CompositeResolver CreateForAot(IFormatterResolver aotResolver)
+        public static IFormatterResolver CreateForAot(IFormatterResolver aotResolver)
         {
-            var composite = new CompositeResolver();
-            composite.RegisterResolver(aotResolver);
-
-            composite.RegisterResolver(
-                BuiltinResolver.Instance,
-                AttributeFormatterResolver.Instance,
-                PrimitiveObjectResolver.Instance);
+            if (aotResolver is null)
+            {
+                throw new ArgumentNullException(nameof(aotResolver));
+            }
 
-            return composite;
+            return new CachingResolver(
+                EmptyFormattersByType,
+                new ReadOnlyCollection<IFormatterResolver>(
+                    new[]
+                    {
+                        aotResolver,
+                        BuiltinResolver.Instance,
+                        AttributeFormatterResolver.Instance,
+                        PrimitiveObjectResolver.Instance,
+                    }));
         }
 
-        /// <summary>
-        /// Adds a resolver to this composite resolver.
-        /// </summary>
-        /// <param name="resolver">The resolver to add.</param>
-        /// <remarks>
-        /// Each registered resolver is appended to the end of the list of resolvers to try for each requested type.
-        /// Registered formatters take precedence over all resolvers.
-        /// </remarks>
-        public void RegisterResolver(IFormatterResolver resolver)
+        public static IFormatterResolver Create(IReadOnlyList<object> formatters, IReadOnlyList<IFormatterResolver> resolvers)
```

Comment:
Why IReadOnlyList<object>?
Maybe, you removed `IMessagePackFormatter` so can't use constraint.

Reply:
> Why `IReadOnlyList<object>`?

Why not? What would you use instead?
`IReadOnlyList<T>` conveys the idea to the caller that the collection must be ordered (as it should, since order of formatters impacts the result) and that we will not change the list within the method. It's an interface so that the caller can provide any sort of ordered collection they want.

> Maybe, you removed `IMessagePackFormatter` so can't use constraint.

Ah, perhaps you're not asking why I use the list type but rather why I use `object` as the generic type argument. That's because this method must accept a list of formatters for many types, and your `IMessagePackFormatter<T>` interface is itself generic, so there is no one interface type that I can accept here without limiting the caller to only providing a formatter for a single type.

Reply:
Ah, I guess a while back yes I must have removed the empty `IMessagePackFormatter` interface. Do you want me to put that back?